support cast decimal to decimal

liukun4515 opened the pull request:
# Which issue does this PR close?

<!---
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

part of https://github.com/apache/arrow-rs/issues/1043

# Rationale for this change
 
 <!---
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?

<!---
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?


<!---
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

alamb commented on arrow/src/compute/kernels/cast.rs:
```diff
@@ -1908,6 +1945,14 @@ mod tests {
+        let input_type = DataType::Decimal(20, 3);
+        let output_type = DataType::Decimal(20, 4);
+        assert!(can_cast_types(&input_type, &output_type));
+        // let array = vec![Some(1123456), Some(2123456), Some(3123456), None];
```
looks like we just need some tests here, and we'll be all set 👍

alamb reviewed (commented):
@liukun4515  I looked at this PR and the code looks good. I think all it needs now is some tests and a rebase and it would be good to go.

I plan to create an arrow 6.5.0 release candidate tomorrow (Thursday) -- let me know if you think we should aim to get this PR merged before doing so.

liukun4515 commented:
> @liukun4515 I looked at this PR and the code looks good. I think all it needs now is some tests and a rebase and it would be good to go.
> 
> I plan to create an arrow 6.5.0 release candidate tomorrow (Thursday) -- let me know if you think we should aim to get this PR merged before doing so.

I rebased this branch with master and added some test cases.
If it looks good for you, you can merge it.
Thank.

alamb commented:
Thanks @liukun4515 ❤️
